libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt

```diff
@@ -13,9 +13,21 @@ fun interface RoutesObserver {
      *
      * The route at index 0, if exist, will be treated as the primary route for 'Active Guidance'.
      *
-     * A list of routes can be modified internally and externally at any time with [MapboxNavigation.setRoutes], or during automatic reroutes, faster route and route refresh operations.
-     *
-     * @param routes list of currently maintained routes
+     * A list of routes can be modified internally and externally at any time with
+     * [MapboxNavigation.setRoutes], or during automatic reroutes, faster route and route refresh operations.
      */
-    fun onRoutesChanged(routes: List<DirectionsRoute>)
+    fun onRoutesChanged(result: RoutesUpdatedResult)
 }
+
+/**
+ * Routes updated result is provided via [RoutesObserver] whenever a list of routes changes.
+ *
+ * The route at index 0, if exist, will be treated as the primary route for 'Active Guidance'.
+ *
+ * @param routes list of currently maintained routes
+ * @param reason why the routes have been updated (re-route, refresh route, and etc.)
+ */
+class RoutesUpdatedResult(
+    val routes: List<DirectionsRoute>,
+    @RoutesExtra.RoutesUpdateReason val reason: String,
```

Comment:
Maybe add `initialLegIndex` here as well? It may be useful and this will also let us remove `initialLegIndex` from `DirectionsSession` interface.

Reply:
we don't broadcast `initialLegIndex` via `RoutesObserver`, and I think we shouldn't 🤔